x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java

```diff
@@ -0,0 +1,217 @@
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardsIterator;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportRequestOptions;
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.transform.action.GetCheckpointAction;
+import org.elasticsearch.xpack.core.transform.action.GetCheckpointAction.Request;
+import org.elasticsearch.xpack.core.transform.action.GetCheckpointAction.Response;
+import org.elasticsearch.xpack.core.transform.action.GetCheckpointNodeAction;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+
+public class TransportGetCheckpointAction extends HandledTransportAction<Request, Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportGetCheckpointAction.class);
+    private final ClusterService clusterService;
+    private final IndicesService indicesService;
+    private final TransportService transportService;
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+
+    @Inject
+    public TransportGetCheckpointAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final IndicesService indicesService,
+        final ClusterService clusterService,
+        final IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(GetCheckpointAction.NAME, transportService, actionFilters, Request::new);
+        this.transportService = transportService;
+        this.indicesService = indicesService;
+        this.clusterService = clusterService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final ClusterState state = clusterService.state();
+        resolveIndicesAndGetCheckpoint(task, request, listener, state);
+    }
+
+    protected void resolveIndicesAndGetCheckpoint(Task task, Request request, ActionListener<Response> listener, final ClusterState state) {
+        String[] concreteIndices = this.indexNameExpressionResolver.concreteIndexNames(
+            state,
+            request.indicesOptions(),
+            true, // includeDataStreams
+            request.indices()
+        );
```

Comment:
This does not work because `GetCheckpointAction` is a cluster action (`cluster:admin/transform/checkpoint`). Authorization does not expand indices for cluster actions, this means the expansion here does not respect user's index access and could expand to anything that the cluster has. The quick fix is likely changing the action from `cluster:...` to `indices:...`.